core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -0,0 +1,1077 @@
+      this.queue = q;
+      this.hasTimeout = hasTimeout;
+      this.timeoutAtNanos = timeoutAtNanos;
+    }
+
+    @Override
+    public boolean block() throws InterruptedException
+    {
+      boolean success = false;
+      if (item != null) {
+        if (hasTimeout) {
+          final long thisTimeoutNanos = timeoutAtNanos - System.nanoTime();
+          if (thisTimeoutNanos < 0) {
+            throw new RE(new TimeoutException("QueuePusher timed out offering data"));
+          }
+          success = queue.offer(item, thisTimeoutNanos, TimeUnit.NANOSECONDS);
+        } else {
+          success = queue.offer(item);
+        }
+        if (success) {
+          item = null;
+        }
+      }
+      return success;
+    }
+
+    @Override
+    public boolean isReleasable()
+    {
+      return item == null;
+    }
+
+    void addItem(E item)
+    {
+      this.item = item;
+    }
+
+    public void offer(E item)
+    {
+      try {
+        addItem(item);
+        ForkJoinPool.managedBlock(this);
+      }
+      catch (InterruptedException e) {
+        throw new RuntimeException("Failed to offer result to output queue", e);
+      }
+    }
+  }
+
+
+  /**
+   * Holder object for an ordered batch of results from a sequence. Batching the results vastly reduces the amount of
+   * blocking that is needed to move results between stages of {@link MergeCombineAction} done in parallel, allowing
+   * the fork join tasks to focus on doing actual work instead of dealing with managed blocking.
+   */
+  static class ResultBatch<E>
+  {
+    static final ResultBatch TERMINAL = new ResultBatch();
+
+    @Nullable
+    private final Queue<E> values;
+    private final boolean isTerminal;
+
+    ResultBatch(int batchSize)
+    {
+      this.values = new ArrayDeque<>(batchSize);
+      this.isTerminal = false;
+    }
+
+    private ResultBatch()
+    {
+      this.values = null;
+      this.isTerminal = true;
+    }
+
+    public void add(E in)
+    {
+      assert values != null;
+      values.offer(in);
+    }
+
+    public E get()
+    {
+      assert values != null;
+      return values.peek();
+    }
+
+    public E next()
+    {
+      assert values != null;
+      return values.poll();
+    }
+
+    boolean isDrained()
+    {
+      return !isTerminal && values.isEmpty();
+    }
+
+    boolean isTerminalResult()
+    {
+      return isTerminal;
+    }
+
+    /**
+     * Convert sequence to yielder that accumulates results into ordered 'batches'
+     */
+    static <E> Yielder<ResultBatch<E>> fromSequence(Sequence<E> sequence, int batchSize)
+    {
+      return sequence.toYielder(
+          new ResultBatch<>(batchSize),
+          new YieldingAccumulator<ResultBatch<E>, E>()
+          {
+            int count = 0;
+            @Override
+            public ResultBatch<E> accumulate(ResultBatch<E> accumulated, E in)
+            {
+              count++;
+              if (count % batchSize == 0) {
+                yield();
+              }
```

Comment:
I can put the yield after the value is added to the batch, if it makes it look less dicey. I'm not sure what a comment would look like? This block seems obvious to me, but maybe I'm too far in and can't properly gauge such things.

Reply:
I guess maybe the comment can explain how `accumulated` is used and what happens when `yield()` is called. For example, it wasn't clear to me how `ResultBatch` is created because `in` is always added to the same `accumulated`. Regarding `yield()`, it wasn't clear what code will be executed when `yield()` is called.